lib-extra/src/main/java/com/diffplug/spotless/extra/config/EclipseConfiguration.java

```diff
@@ -0,0 +1,257 @@
+		} catch (MalformedURLException e) {
+			/*
+			 *  This exception must be prevented by the strict syntax checking of
+			 *  SemanticVersion and unit-tests of all allowed versions.
+			 */
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * Modify default Maven dependencies configured for the formatter.
+	 * The default dependencies are located in {@link #ECLIPSE_FORMATTER_RESOURCES}.
+	 */
+	public void setDepenencies(String... coordinates) {
+		coordinateAdaptations = new MavenCoordinates();
+		coordinateAdaptations.update(coordinates);
+		invalidateState();
+	}
+
+	/** Set settings files containing Eclipse preferences */
+	public void setPreferences(Iterable<File> settingsFiles) {
+		this.settingsFiles = settingsFiles;
+		invalidateState();
+	}
+
+	/** The current state becomes invalid in case a configuration item has successfully changed */
+	private void invalidateState() {
+		state = null;
+	}
+
+	/** Creates the state of the configuration */
+	public EclipseConfiguration.State get() {
+		if (null == state) {
+			state = new State(
+					jarProvisioner,
+					createMavenCoordinates(),
+					version,
+					settingsFiles);
+		}
+		return state;
+	}
+
+	private MavenCoordinates createMavenCoordinates() {
+		MavenCoordinates coordinates = new MavenCoordinates();
+		coordinates.add(defaultCoordinatesUrl);
+		coordinates.update(coordinateAdaptations);
+		return coordinates;
+	}
+
+	/**
+	 * State of Eclipse configuration items, providing functionality to derived information
+	 * based on the state.
+	 */
+	public static class State implements Serializable {
+		// Not used, only the serialization output is required to determine whether the object has changed
+		private static final long serialVersionUID = 1L;
+
+		private final FileSignature settingsFiles;
+		private final MavenCoordinates coordinates;
+		private final SemanticVersion version;
+
+		/*
+		 * Fields are transient because not needed to uniquely identify a Eclipse configuration state, and also because
+		 * Gradle only needs this class to be serializable so it can compare its members for incremental builds.
+		 */
+		private transient ClassLoader lazyClassLoader;
+		@SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+		private final transient Provisioner jarProvisioner;
+
+		/** State constructor expects that all passed items are not modified afterwards */
+		protected State(Provisioner jarProvisioner, MavenCoordinates coordinates, SemanticVersion version, Iterable<File> settingsFiles) {
+			this.jarProvisioner = jarProvisioner;
+			this.coordinates = coordinates;
+			this.version = version;
+			lazyClassLoader = null;
+
+			try {
+				this.settingsFiles = FileSignature.signAsList(settingsFiles);
+			} catch (NullPointerException e) {
+				throw new UserArgumentException(null, "Configuration file settings", e);
+			} catch (IOException e) {
+				throw new RuntimeException(e); //Canonical path problems are not user argument problems
+			}
+		}
+
```

Comment:
I admit that I just added the functionality because it is demanded by the `SerializableEqualityTester`. Later I found that actually the `FileSignature` seems to have no test for serialization.
So I was not sure anymore what I really need and decided to wait for your review.
I really would like to have a test. Is there already a test utility similar to  the `SerializableEqualityTester` omitting the hash test? Or shall I provide one?

Reply:
The `SemanticVersion` shall provide three things:
- Be convenient for the user so it treats for example `4.5` and `4.5.0` equally.
- Check user input on a high level so that useful responses are provided. If later on some method reports that jars:/<some_path>/<version> has not been found, it is more difficult for the user to see that just his version configuration is a problem. This is especially a problem when just minor things, like a white space, are incorrect.
- Allow simple comparison (less than, greater than) so that the Eclipse formatter step can determine whether it needs some different interface implementation.

Reply:
Beside the convenience / checks for user data, the `MavenCoordinates` shall
- Distinguish whether default settings are added
- Versions are updated
- Provide the coordinates as a set to the provisioner to allow transitive dependency mediation. To my understanding the concepts here in Maven and Gradle slightly differ. Keep in mind that the default versions are fixed, but I expect users (and me as a developer) to use version ranges to do some testings.

Reply:
> so it treats for example 4.5 and 4.5.0 equally.

This won't work as you intend for up-to-date checking.  Up-to-date checking is based *solely* on the serialized representation, it doesn't care about `.equals()`.

> I added the functionality because it is demanded by the SerializableEqualityTester

SerializableEqualityTester is only meant to be used on either steps or the entire step state.  Not the *components* of step state.

Up-to-date checking works like this:

- every step is a subclass of this, which means that equals and hashCode are implemented in terms of the state's serialized representation
https://github.com/diffplug/spotless/blob/c6c32e3a716bdcc07393f51266b9765fdf7db0c2/lib/src/main/java/com/diffplug/spotless/FormatterStep.java#L65-L78
- if the state has sub-components (such as FileSignature or JarState), they get serialized into the state and compared there.  FileSignature and JarState don't declare an equals or hashCode, because it would never run.  Equality testing happens at the *entire state* level, not component-by-component.

> SemanticVersion and MavenCoordinates

Provisioner only takes a List<String>.  So I don't see how we will pass this detailed semantic information to the step's implementation.  We can always do feature testing a-la:

https://github.com/diffplug/spotless/blob/c6c32e3a716bdcc07393f51266b9765fdf7db0c2/lib/src/main/java/com/diffplug/spotless/scala/ScalaFmtStep.java#L98-L115

Reply:
Thanks for the hint with the dependency lock files generated by Gradle. Did not know about the feature, but this is exactly what I need to generate initial default locks for the Eclipse formatter implementations. So I will modify the current solution to cope with these files.
The workflow for providing new Eclipse formatter implementation version will in that case be:
- Go to `_ext` and call a build
- If all tests are passing with the new versions and additional tests prove that the new formatting feature (what ever the reason for the update is) is applied, generate the locks and make it available in `extra` 
- If the tests failing with the new Eclipse dependency versions, a new version of the formatter implementation is required.But I hope, with some more experience, that the framework and the formatter adaptations will once be generic enough so that a modification will seldom required.

Reply:
>>   so it treats for example 4.5 and 4.5.0 equally.

> This won't work as you intend for up-to-date checking.

That's why 4.5 directly is converted to 4.5.0. So it should work.

> if the state has sub-components (such as FileSignature or JarState), they get serialized into the state...

This is now understood. Thanks. 

> We can always do feature testing a-la

Sorry, you lost me. Are we talking about testing serialization? I though that I would need to generate an `ObjectInputStream` for testing, to see that it really works, or is this guaranteed already by the compiler? Never worked much with serialization, so I was so keen to have some testing...

Reply:
Nope, I mean that instead of saying "if version < 3 do this, else do that", you can instead use reflection to test what methods are present (as in the ScalaFmtStep example).  I'm making the case for why we shouldn't have SemanticVersion and MavenCoordinate when we can just have `List<String>` instead.

Reply:
I am aware of the principal, but to be honest, I am quite reluctant to test which package name to use when I know that with a certain version I introduced had some refactoring. One key point of this refactoring is to avoid this ugly "deprecation flagging" we had in EclipseFormatterStep.

Reply:
Maybe I should explain my idea better:

For every formatter step implementation change, we modified methods (e.g. create) for all Eclipse formatters and declared the old once as deprecated.
Hence, I would prefer to pass an `EclipseConfiguration` item which is used by the Formatter steps like a POJO. No deprecation if I add more config items.... `SemanticVersion` and 
`MavenCoordinates` are explicitly not public. They provide functionality which I consider in itself encapsulated and not Eclipse specific. Since they are not public, I see no need at all to keep their non-public methods backward compatible in case I find it useful (for Eclipse or non-Eclipse formatters) to make their functionality depend on additional parameters, or handle different sources.

Furthermore you asked me, when I introduced the `GrEclipseFormatterStep`, too copy the internal State class approach so that in case of interface changes, both Eclipse formatters are independent. Till now, this only caused a deprecation of methods in both Eclipse formatters.
The new approach splits the state and method provider on purpose. It allows common code (we will have some more Eclipse formatters soon ...) without limiting the flexibility of individual interface evolution.

When it comes to interface changes, the differences can be handled quite cleanly by the Eclipse FormatterStep (method provider) members. Only on drastic changes, the method provider must support different  types of methods. 

Changes are expected to be cascading. Example:
*  Since version X, we use a new package name
*  Since version Y (Y > X), we also pass the file, because somewhere in the Eclipse plugin internals its needed

Hence I consider "feature testing" an ugly solution for the problem.

Reply:
You make a good argument against feature testing.  For package-private, we definitely have a lot of leeway.

Do you think something like this could work?

```java
if (versionOf(deps, 'org.eclipse.jdt').isLessThan("4.7")) {
   // use such and such

static Version versionOf(List<String> deps, String target) {
   for (String dep : deps) {
      if (dep.startsWith(target + ":)) {
        return new Version(dep.substring(dep.lastIndexOf(':')));
     }
}
```

My concern is that it's quite a lot of code.  Each LOC makes it harder for new contributors to join and fix things.  It's package-private, so at the end of the day if you need it then we can have it.  But the smaller it gets, the more help we'll get (and the less help we'll need).

Reply:
I am not sure what your goal is. I can reduce the LOC by integrating the functionality back into EclipseConfiguration, though I would still prefer to have it a separated inner class, to keep the code readable.
Your proposal still seem to assume that the classes are used externally, but they shall explicitly not, since this, as you pointed out and I a fully agree, would make of the developer harder. My goal for the Formatter Steps is something like this (the `versionHigher` does not exist yet):

```java
class EclipseJdtFormatterStep .... {
...
private static String  INITV = "4.7.2"; //Highest version compatible to initial interface
private static VERSIONS[] = {"4.6.1", "4.6.3", "4.7.0", "4.7.1", INITV, "4.7.3"};
private final String implClassPath;

	public static FormatterStep createStep(EclipseConfiguration config) {
                String classPath = config.versionHigher(INITV)? "com.diffplug.spotless.extra.java.eclipse.EclipseJdtFormatterStepImpl" : ""com.diffplug.gradle.spotless.java.eclipse.EclipseFormatterStepImpl"
		return FormatterStep.createLazy(NAME, config, new EclipseJdtFormatterStep(classPath));
	}

}
```

The EclipseConfiguration and maybe later on some other XyzConfgiuration shall be the front end. The other package scope classes are just generic tools.

Reply:
The tools have three goals:
*  Sanitize user input where convenient (3.4 -> 3.4.0, trim input, ...)
*  Validate user input if easily possible (for example the version range syntax is too complex, so I don't check it)
*  Provide functionality for the step configuration class to gain the information it needs (like comparison, initializing from specific file format, ...).

Reply:
One remaining question about the `SerializableEqualityTester`: 
Do we have something like a `SerializableTester`? Do you think it is worth to have it, or are the compiler checks good enough?

Reply:
I think the compiler check is good enough.